### PR TITLE
Refactored TrustChain so it's a single ledger now

### PR DESCRIPTION
**Pull request by devos50**
In this PR, I refactored the TrustChain overlay so it's a generic ledger that is reusable across several domains. The most important change is that each transaction now should have a type (i.e. `ask`, `tribler_bandwidth`, `attestation` etc). This transaction type is also stored in the database as a separate column for easy lookups. For instance, this allows us to get all blocks with an `attestation` transaction.

In addition, I also added a timestamp to each TrustChain record. These changes are included in this PR since it also breaks backward compatibility so I guess it's ok to combine it with the other breaking change (merging all chains together).

### CODE REVIEW

**Comment by qstokkink** [comment]
Could we keep this on hold for a few days until I stabilize the attestation community again?

**Comment by devos50** [comment]
Sure, no problem!

File: ipv8/attestation/trustchain/community.py [hunk 1]
```diff
@@ -60,6 +59,9 @@ def __init__(self, *args, **kwargs):
         self.logger.debug("The trustchain community started with Public Key: %s",
                           self.my_peer.public_key.key_to_bin().encode("hex"))
         self.broadcast_block = True  # Whether we broadcast a full block after constructing it
+        self.listeners_map = {}  # Map of block_type -> [callbacks]
```
Comment (qstokkink) [inline ipv8/attestation/trustchain/community.py]: Could you refactor the listeners into a single `Listener` class? The listener object, `should_sign` and `block_class` should all be connected anyway. This is starting to look like the TunnelCommunity.

Comment (devos50) [reply ipv8/attestation/trustchain/community.py]: @qstokkink the `block_class` and `Listener` are not necessarily connected, they can be different objects.

Also, there can be multiple listeners for one block type but there's at most one `block_class` for each type. I don't see how we could generalize this nicely. Do you have any suggestions?

Comment (qstokkink) [reply ipv8/attestation/trustchain/community.py]: I don't see how a listener could possibly comprehend what it is called back with, if it does not know the block class it is getting called back with. `self.notify_listeners(block)` gives a block of type `block = self.get_block_class(block_type)`.

A Listener could be something like:

```python
class TrustChainBlock(SomeBlockSuperclass):
    TYPE = 'bandwidth!'

class MyBeautifulListener(SomeListenerSuperclass):
    BLOCK_CLASS = TrustChainBlock

    def should_sign(block):
        return True
```

File: ipv8/attestation/trustchain/community.py [hunk 2]
```diff
@@ -230,10 +284,6 @@ def process_half_block(self, blk, peer):
         self.logger.debug("Block validation result %s, %s, (%s)", validation[0], validation[1], blk)
         if validation[0] == ValidationResult.invalid:
             return
-        elif not self.persistence.contains(blk):
-            self.persistence.add_block(blk)
```
Comment (qstokkink) [inline ipv8/attestation/trustchain/community.py]: Why was this removed?

Comment (devos50) [reply ipv8/attestation/trustchain/community.py]: Because the block is already added to the database in the `validate_persist_block` method if it is valid. These two lines I removed here were duplicate and unnecessary.

**Comment by devos50** [comment]
@qstokkink I'm not sure why it is necessary to impose shutdown constraints on the listeners/TrustChain community. I think it is sufficient to simply not notify the listeners when the TrustChain community is shutting down or has finished doing so. I've implemented it like this for now (by keeping track of a `shutting_down` variable in the community class). Please let me know whether you agree or not 👍

**Comment by ichorid** [comment]
(Disclaimer: I'm still not familiar with how TrustChain operates, so I am not sure about this one.)

@qstokkink, I wonder how this change would affect your judgment about the potential TrustChain replication problem. I guess, it was you who told me that the typical TrustChain for channel voting is only few hundred kilobytes, so it is easy for counterparties to copy it over completely, store it locally and check it.
I've checked TrustChain DB sizes for tunnel helpers on our oldest hose: some files are over 500 MB in size. If all the data is in the same domain now, this means the counterparty would have to download the whole chain including all records on traffic exchange, etc. 
Am I right?

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
@qstokkink I've updated and improved the TrustChain documentation in a separate commit after your review.

File: ipv8/attestation/trustchain/community.py [hunk 3]
```diff
@@ -39,18 +39,26 @@ class TrustChainCommunity(Community):
...
-                        "a2961b003affd2964f92c457572f2f79de0968c42698e2d1cfb371dd71b275332a0a4c19f35f16166272baae8e23"
-                        "0bba377cc5c40643b83206088075559ec2f13a090e8786d04d84802268bef12e52983978da360589a2b7e293ce4f"
-                        "16d02f37da2c3256f4703b9623d3750f7af437befebc8935c0f0726f58c1c1e9").decode("HEX"))
+    master_peer = None  # Determined at runtime
 
     def __init__(self, *args, **kwargs):
         working_directory = kwargs.pop('working_directory', '')
         db_name = kwargs.pop('db_name', self.DB_NAME)
+        self.testnet = kwargs.pop('testnet', False)
+        TrustChainCommunity.master_peer = self.TESTNET_PEER if self.testnet else self.MAINNET_PEER
```
Comment (ichorid) [inline ipv8/attestation/trustchain/community.py]: @devos50 Could you please instead do it in a more generalized way, like I did with `bootstrap_override` option? It would make testing much more convenient for other members of the team.

File: ipv8/test/attestation/trustchain/test_block.py [hunk 4]
```diff
@@ -14,25 +14,25 @@ class TestBlock(TrustChainBlock):
...
                 self.key = key
             else:
                 self.key = crypto.generate_key(u"curve25519")
 
-            TrustChainBlock.__init__(self, (
-                encode(transaction), self.key.pub().key_to_bin(), random.randint(50, 100), other, 0,
-                sha256(str(random.randint(0, 100000))).digest(), 0, 0))
+            TrustChainBlock.__init__(self, (block_type,
+                                            encode(transaction), self.key.pub().key_to_bin(), random.randint(50, 100),
+                                            other, 0, sha256(str(random.randint(0, 100000))).digest(), 0, 0, 0))
```
Comment (ichorid) [inline ipv8/test/attestation/trustchain/test_block.py]: Formatting? Do all these spaces really increase readability?

File: ipv8/attestation/trustchain/payload.py [hunk 5]
```diff
@@ -237,16 +264,17 @@ class HalfBlockPairBroadcastPayload(HalfBlockPairPayload):
...
+                 link_sequence_number2, previous_hash2, signature2, block_type2, transaction2, timestamp2, ttl):
         super(HalfBlockPairBroadcastPayload, self).__init__(public_key1, sequence_number1, link_public_key1,
                                                             link_sequence_number1, previous_hash1, signature1,
-                                                            transaction1, public_key2, sequence_number2,
-                                                            link_public_key2, link_sequence_number2, previous_hash2,
-                                                            signature2, transaction2)
+                                                            block_type1, transaction1, timestamp1, public_key2,
+                                                            sequence_number2, link_public_key2, link_sequence_number2,
+                                                            previous_hash2, signature2, block_type2, transaction2,
+                                                            timestamp2)
```
Comment (ichorid) [inline ipv8/attestation/trustchain/payload.py]: Using these many unnamed arguments with basically no formatting is asking for troubles...

File: ipv8/attestation/trustchain/payload.py [hunk 6]
```diff
@@ -168,27 +183,31 @@ class HalfBlockPairPayload(Payload):
...
     """
 
-    format_list = ['74s', 'I', '74s', 'I', '32s', '64s', 'varlenI'] * 2
+    format_list = ['74s', 'I', '74s', 'I', '32s', '64s', 'varlenI', 'varlenI', 'Q'] * 2
 
     def __init__(self, public_key1, sequence_number1, link_public_key1, link_sequence_number1, previous_hash1,
-                 signature1, transaction1, public_key2, sequence_number2, link_public_key2, link_sequence_number2,
-                 previous_hash2, signature2, transaction2):
+                 signature1, block_type1, transaction1, timestamp1, public_key2, sequence_number2, link_public_key2,
+                 link_sequence_number2, previous_hash2, signature2, block_type2, transaction2, timestamp2):
```
Comment (ichorid) [inline ipv8/attestation/trustchain/payload.py]: That's a lot of unnamed arguments. Thank God we don't have to worry if we misplace one of them because Python will figure out the type conversions :see_no_evil:

File: ipv8/attestation/trustchain/payload.py [hunk 7]
```diff
@@ -77,12 +83,13 @@ class HalfBlockBroadcastPayload(HalfBlockPayload):
...
-    format_list = ['74s', 'I', '74s', 'I', '32s', '64s', 'varlenI', 'I']
+    format_list = ['74s', 'I', '74s', 'I', '32s', '64s', 'varlenI', 'varlenI', 'Q', 'I']
 
     def __init__(self, public_key, sequence_number, link_public_key, link_sequence_number, previous_hash,
-                 signature, transaction, ttl):
+                 signature, block_type, transaction, timestamp, ttl):
         super(HalfBlockBroadcastPayload, self).__init__(public_key, sequence_number, link_public_key,
-                                                        link_sequence_number, previous_hash, signature, transaction)
+                                                        link_sequence_number, previous_hash, signature,
+                                                        block_type, transaction, timestamp)
```
Comment (ichorid) [inline ipv8/attestation/trustchain/payload.py]: A lot of unnamed arguments.

File: ipv8/attestation/trustchain/block.py [hunk 8]
```diff
@@ -34,12 +35,15 @@ def __init__(self, data=None, serializer=Serializer()):
...
             self.insert_time = None
         else:
-            _, self.transaction = decode(str(data[0]))
-            (self.public_key, self.sequence_number, self.link_public_key, self.link_sequence_number, self.previous_hash,
-             self.signature, self.insert_time) = (data[1], data[2], data[3], data[4], data[5], data[6], data[7])
+            _, self.transaction = decode(str(data[1]))
+            (self.type, self.public_key, self.sequence_number, self.link_public_key, self.link_sequence_number,
+             self.previous_hash, self.signature, self.timestamp, self.insert_time) = (data[0], data[2], data[3],
+                                                                                      data[4], data[5], data[6],
+                                                                                      data[7], data[8], data[9])
```
Comment (ichorid) [inline ipv8/attestation/trustchain/block.py]: I love that formatting! :+1:

File: doc/trustchain.md [hunk 9]
```diff
@@ -1,18 +1,58 @@
...
+| --- | --- | --- |
+| _type_ | string | The type of the block. Can be any arbitrary string. |
+| _transaction_ | dictionary | The transaction that this block describes. |
+| _public_key_ | string | The public key in binary format of the block creator. |
+| _link_public_key_ | string | The public key of the transaction counterparty. |
+| _sequence_number_ | integer | The sequence number of the block. The sequence number of the genesis block is 1. |
+| _link_sequence_number_ | integer | The sequence number of the linked block. |
+| _previous_hash_ | string | The SHA256 hash of the previous block in the chain. |
+| _signature_ | string | The signature of the block. |
+| _timestamp_ | integer | The epoch timestamp defining when this block is created (with milliseconds precision). |
```
Comment (ichorid) [inline doc/trustchain.md]: Thumbs up for documentation! :1st_place_medal:

File: ipv8/attestation/trustchain/community.py (hunk 3 above)
Comment (devos50) [reply ipv8/attestation/trustchain/community.py]: I'm not sure whether it is desirable for the TrustChain community to generalise which master peer we are using (i.e. make it possible to change the master peer from the IPv8 configuration dictionary). The master peer is rarely changed, except when we push some changes that are not backward compatible (like this PR).

In this case, I argue that it is helpful to make a distinction between a mainnet and testnet. We should use the testnet when running integration tests and experiments. Besides this mainnet/testnet, I do not see another use-case for changing the master peer and such, I don't think it is necessary to generalize this.

File: ipv8/test/attestation/trustchain/test_block.py (hunk 4 above)
Comment (devos50) [reply ipv8/test/attestation/trustchain/test_block.py]: I think it increases readability and it follows [the PEP8 conventions](https://www.python.org/dev/peps/pep-0008/).

File: ipv8/attestation/trustchain/payload.py (hunk 5 above)
Comment (devos50) [reply ipv8/attestation/trustchain/payload.py]: I agree. Payloads which contain much data have many arguments. Maybe we can put each argument on a new line?

File: ipv8/attestation/trustchain/payload.py (hunk 6 above)
Comment (devos50) [reply ipv8/attestation/trustchain/payload.py]: In Python, we have to rely on proper unit tests to make sure that we do not make mistakes when passing arguments.

**Comment by devos50** [comment]
@ichorid I have some comments on your comments 👍

File: ipv8/attestation/trustchain/payload.py (hunk 6 above)
Comment (ichorid) [reply ipv8/attestation/trustchain/payload.py]: :crying_cat_face:

File: ipv8/test/attestation/trustchain/test_block.py (hunk 4 above)
Comment (ichorid) [reply ipv8/test/attestation/trustchain/test_block.py]: If it fits in 80 lines that's OK.

File: ipv8/attestation/trustchain/payload.py (hunk 5 above)
Comment (ichorid) [reply ipv8/attestation/trustchain/payload.py]: The problem is not in the formatting, but in the number of arguments and potential pitfall of mis-assigning them. I would prefer an intermediate object, or dict, or named arguments instead.
If you think it's OK to leave it as it is, that's up to you. Its only a minor note on style, nothing serious.

**Comment by qstokkink** [comment]
retest this please

**Comment by devos50** [comment]
retest this please
